pkg/blockbuilder/blockbuilder.go

```diff
@@ -439,6 +449,11 @@ func (b *BlockBuilder) consumePartition(
 			}
 			lastRec = rec
 		}
+		if lag <= 0 {
+			// TODO(codesome): Verify with Vladimir if this is okay. This makes tests go faster because
+			// otherwise the fetch waits until timeout for the next one.
+			break
+		}
```

Comment:
I'm not too comfortable with this one:

1. We now have two static end-of-cycle markers: `lag` and `cycleEnd`. I see the point about making tests faster, but my concern is that those guards can cause us to miss records in real scenario, when writes and reads to Kafka happen concurrently.
1. In practice, the `lag` will _almost_ always be positive here because an active partition will always have more records in the backlog, than what we consume in one cycle. I.e. distributors will keep producing records when (and after) we enter `nextConsumeCycle`.